File: ch2/Binary_Search/algorithm.cpp
#include <vector>
template <class T>
int binary_search(std::vector<T> L, T X) {
    int low = 0;
    int high = L.size() - 1;
    while (low <= high) {
        int mid = (low+high)/2;
        if (L[mid] == X) {
            return mid;
        } else if (L[mid] > X) {
            high = mid - 1;
        } else if (L[mid] < X) {
            low = mid + 1;
        }
    }
    return -1;
}

File: ch2/Binary_Search/algorithm.py
def binary_search(L, X):
    low = 0
    high = len(L)-1
    while low <= high:                          #this fails when we have one element so low and high are equals
        mid = (low+high)//2                    #and then we moved low one step up or high one step down
        if L[mid] == X:
            return mid
        elif X>L[mid]:
            low = mid + 1
        elif X<L[mid]:
            high = mid - 1
    return -1

L = [1,4,7,9,10,20,25,34,90,10]
print(binary_search(L, 25))
